### PR TITLE
Diaganostics and metrics clean up

### PR DESCRIPTION
**Pull request by JamesNK**
Some clean-up from https://github.com/dotnet/aspnetcore/pull/46834

Made `IHttpActivityFeature` and `IHttpMetricsTagsFeature` behave consistently. The "owner" of the features responsible for setting them is hosting. This allows us to have a source of truth implementation which means we can access the underlying `List<T>` on metrics tags directly instead of going through the feature interface.

Previously hosting would check if the server had an activity feature. Now, if the server sets the feature, it is overridden by hosting. This was probably done for testing. It's a change but very, very unlikely anyone would do this (implement a new server and choose to override `IHttpActivityFeature`).

### CODE REVIEW

File: src/Hosting/Hosting/test/HostingApplicationTests.cs [hunk 1]
```diff
@@ -221,8 +257,7 @@ private class TestHttpActivityFeature : IHttpActivityFeature
     }
 
     [Fact]
-    [QuarantinedTest("https://github.com/dotnet/aspnetcore/issues/38736")]
```
Comment (JamesNK) [inline src/Hosting/Hosting/test/HostingApplicationTests.cs:224]: This quarantined test has been removed.

File: src/Hosting/Hosting/test/HostingApplicationTests.cs [hunk 2]
```diff
@@ -204,7 +238,9 @@ public void IHttpActivityFeatureIsPopulated()
         var initialActivity = Activity.Current;
 
         // Create nested dummy Activity
-        using var _ = dummySource.StartActivity("DummyActivity");
+        using var dummyActivity = dummySource.StartActivity("DummyActivity");
+        Assert.NotNull(dummyActivity);
+        Assert.Equal(Activity.Current, dummyActivity);
```
Comment (JamesNK) [inline src/Hosting/Hosting/test/HostingApplicationTests.cs:243]: Attempting to narrow down the cause of this flaky test.

**Comment by mitchdenny** [comment]
Looks good so far, but will wait to approve until you've finished running down the flakey test.

**Comment by JamesNK** [comment]
> Looks good so far, but will wait to approve until you've finished running down the flakey test.

Don't need to hold up merging this.

`IHttpActivityFeatureIsPopulated` hasn't failed in the last 30 days. I'm adding extra logging/asserts (already done) and then marking the issue as fixed. If there are failures after a month then it can be unqualified.

**Review by mitchdenny — APPROVED** [review]

**Pull request merged**